python/ffsim/tenpy/circuits/lucj_circuit.py

```diff
@@ -0,0 +1,102 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+from __future__ import annotations
+
+import numpy as np
+from tenpy.algorithms.tebd import TEBDEngine
+from tenpy.networks.mps import MPS
+
+import ffsim
+from ffsim.tenpy.circuits.gates import (
+    apply_diag_coulomb_evolution,
+    apply_orbital_rotation,
+)
+from ffsim.tenpy.util import product_state_as_mps
+from ffsim.variational.ucj_spin_balanced import UCJOpSpinBalanced
+
+
+def lucj_circuit_as_mps(
+    norb: int,
+    nelec: int | tuple[int, int],
+    ucj_op: UCJOpSpinBalanced,
+    options: dict,
+    *,
+    norm_tol: float = 1e-5,
+) -> tuple[MPS, list[int]]:
```

Comment:
To be more consistent with the rest of ffsim's API, this function should have this API:
```python
def lucj_circuit_as_mps(
    mps: MPS,
    ucj_op: UCJOpSpinBalanced,
    norb: int,
    nelec: int | tuple[int, int],
    options: dict,
    *,
    norm_tol: float = 1e-5,
) -> MPS:
```
I think the output should just be the MPS. If we want to get the bond dimensions, we should figure out a way to extract that from the engine separately. Maybe we can use a callback function.

Reply:
Yes, I think the most important piece of information from `chi_list` is the maximum bond dimension reached during the simulation, and this you can see by simply looking at the bond dimensions of the resulting mps. The added value of `chi_list` is quite niche, since it can tell you specifically at which points in the circuit the bond dimension grows and by how much. I can remove it for now.

Reply:
Another option is that I create a `CircuitMPS` class that inherits from `MPS` and has some additional attributes, such as `chi_list`.

Reply:
I think that we want to have something like `chi_list` because it tells us explicitly when the bond dimension is saturated, rather than relying solely on observables, such as energy convergence. But I'm not sure how best to do it...

Reply:
since `eng` stores a reference to the MPS, can we just pass that as input to this function? See my other comment. Also, btw I updated the first post of this thread, I meant to suggest a new name for this function, `apply_ucj_op_spin_balanced`.

Reply:
Yes, we can just pass `eng`. We can simply use `eng._resume_psi` to get `psi`, if needed.